Blog post: Temporal's Access Control Strategy in AWS (#1215)
temporalio/documentation: blog/deep-dive-temporal-access-control-strategy-aws.md

---
slug: deep-dive-temporal-access-control-strategy-aws
title: A Deep Dive into Temporal's Access Control Strategy in AWS
author: Brandon Sherman
author_title: Staff Security Engineer
author_image_url: /img/brandon-sherman.png
tags:
  - security
  - transparency
  - cloud
date: 2022-06-06T00:00:00Z
image: /img/temporal-access-control.jpeg
---

This blog post gives some insight into Temporal's strategy for securing our cloud environment.

![Photo of dodgy-looking ATM beside a security grate](/img/temporal-access-control.jpeg)

<!-- truncate -->

> It also calls attention to an unexpected facet of AWS access policies encountered along the way. Finally, I'll come around to what I discovered—what I thought I wanted isn't what I needed—and describe what I need.

Author: Thanks for highlighting this. I addressed both this and the following suggestion by moving the first sentence above the image.

> 

Author: Good call; added.

> Our long-term goal is to develop an access control mechanism that can be flexible, scalable, and secure. After we reviewed various strategies, a [Biba-style model](https://en.wikipedia.org/wiki/Biba_Model) of segmenting our AWS accounts seems to fit our needs the best. Each account forms a hard<sup>1</sup> "perimeter" and provides an easy way to set, measure, and maintain the maximum "blast radius" of each account. If we assume an account has been breached, what is the maximum extent of damage an attacker can inflict? 

Author: I accepted the first and third edits, but using "reviewing" creates a dangler; that is, what is the actor doing the reviewing?

Setting up numerous smaller accounts helps to constrain possible damage, but this approach would also make our management infrastructure a very compelling target. If only one management account has permissions configured to allow those with access to assume roles into all other accounts, and the management account is breached, the blast radius goes from being extremely small to being extremely large. **Clearly, this is not ideal.**

We want to provide strong isolation guarantees to our customers as well as ourselves. An engineer paged at 3:00 AM needs to be comfortable in the knowledge they will not cause large-scale problems when they fix the isolated problem they were paged about. To provide these strong isolation guarantees, we plan to have a small number of trusted core accounts. Think of these as Ring 0: interacting with the infrastructure of these accounts in any way should be a big event.<sup>2</sup> The next ring of accounts, Ring 1, will only listen to commands from Ring 0 and issue commands to Ring 2. Similarly, our core accounts—those Ring 0 and Ring 1 accounts—should not be able to read the data in those outer rings.

> It's easy enough to describe this approach, but promises are meaningful only if you follow through with them! The first step for me was to codify this into our AWS service control policies (SCPs). SCPs are extremely powerful tools because they apply to all principals within an AWS account, including the all-powerful root user. Effective SCP layout and implementation is critical to a well-secured AWS environment.

Reviewer: ```suggestion
It's easy enough to describe this approach, but promises are only meaningful if you follow through with them! The first step was to codify this concept into our AWS service control policies (SCPs). SCPs are extremely powerful tools because they apply to all principals within an AWS account, including the all-powerful root user. Effective SCP layout and implementation is critical to a well-secured AWS environment.
```

## Digression 

> Up until SCPs became available, all AWS Identity Access Management (IAM) functioned via [discretionary access control](https://en.wikipedia.org/wiki/Discretionary_access_control). There was no way to restrict the actions of an AWS account. Note that this differs from restricting the actions of an IAM User account within the AWS account. Controlling the permissions of subsections of an AWS account has been possible since the release of AWS IAM, but these were entirely the user's responsibility to put in place—and it was impossible to mandate that some actions always be denied.<sup>3</sup> With privilege escalation readily usable—for either legitimate or malicious purposes—within an account, it remains unwise to rely on a single AWS account.

Reviewer: ```suggestion
Prior to SCPs, all AWS Identity Access Management (IAM) functioned via [discretionary access control](https://en.wikipedia.org/wiki/Discretionary_access_control). There was no way to restrict the actions of an AWS account. Controlling the permissions of subsections of an AWS account has been possible since the release of AWS IAM, but these were the user's responsibility to put in place—and it was impossible to mandate that some actions always be denied.<sup>3</sup> With privilege escalation readily usable—for either legitimate or malicious purposes—within an account, it remains unwise to rely on a single AWS account.
```

> When SCPs were released, AWS IAM gained the ability to set [mandatory access controls](https://en.wikipedia.org/wiki/Mandatory_access_control). An SCP applied to anything and everything in the AWS account, including the root user. It doesn't matter if the permissions are allowed within the account—SCPs are more powerful. Now, to escalate permissions within an AWS account, an attacker must first break into another AWS account and take actions to enable their attack on the initial target. AWS CloudTrail logs an event any time a user encounters an "Access denied" message due to SCPs, and CloudTrail can also log actions such as moving accounts between OUs and adding, removing, or modifying SCPs for an OU or account. In short, breaking into an AWS account this way requires a series of steps that are much noisier–and therefore more easily monitored.

Reviewer: ```suggestion
When SCPs were released, AWS IAM gained the ability to set [mandatory access controls](https://en.wikipedia.org/wiki/Mandatory_access_control). An SCP applied to anything and everything in the AWS account, including the root user. It doesn't matter if the permissions are allowed within the account. SCPs are more powerful. 

To escalate permissions within an AWS account, an attacker must first break into another AWS account and take actions to enable their attack on the initial target. AWS CloudTrail logs an event any time a user encounters an "Access denied" message due to SCPs, and CloudTrail can also log actions such as moving accounts between OUs and adding, removing, or modifying SCPs for an OU or account. In short, breaking into an AWS account this way requires a series of steps that are much noisier, making them easier to monitor.
```

## How Mandatory are SCPs?

> The stage is set. We have a group of AWS accounts, some of which need the ability to assume a role into a separate group of accounts to perform infrastructure/maintenance/control tasks. And we have the ability to **mandate** what an account is able to do. All that I should need to do would be to mandate that our accounts operate in this manner. We will have our rings of AWS accounts, and AWS will handle the enforcement of what a principal can do. This is how AWS recommends SCPs be used, as guardrails around an account.<sup>4</sup>

Reviewer: ```suggestion
Let's set the stage. 

We have a group of AWS accounts. Some accounts need the ability to assume a role into a separate group of accounts to perform infrastructure/maintenance/control tasks.

We have the ability to **mandate** what an account is able to do. All that I need to do is mandate that our accounts operate in this manner. We will have our rings of AWS accounts, and AWS will handle the enforcement of what a principal can do. This is how AWS recommends SCPs be used, as guardrails around an account.<sup>4</sup>
```

> I wasn't worried about the positive case of granting the permissions of one account to assume-role into another. Instead, I wanted to test the negative case: disallowing any principal except for a particular role in an account in the correct ring to assume-role.<sup>5</sup>

Reviewer: ```suggestion
I wasn't worried about the positive case: granting the permissions of one account to assume-role into another. Instead, I wanted to test the negative case: disallowing any principal except for a particular role in an account in the correct ring to assume-role.<sup>5</sup>
```

I reached for the tool that I assumed would be the right one: an SCP with conditionals.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Deny",
      "Action": [
        "sts:AssumeRole"
      ],
      "Resource": [
        "arn:aws:iam::575555303850:role/TestAssumedRole"
      ],
      "Condition": {
        "StringNotEqualsIgnoreCase": {
          "aws:PrincipalTag/AssumeRole": [
            "OK"
          ]
        }
      }
    }
  ]
}
```

My understanding was that this SCP would disallow the role in this account (creatively named `TestAssumedRole`) from being assumed, unless the Principal performing the AssumeRole API call had a key/value tag pairing of "AssumeRole/OK". The initial test was successful. **But when I removed the tag from the role doing the assumption ... it still worked.**

> Surprising, right? Well, maybe not. IAM is complicated. My initial assumption was that I had somehow screwed up the Condition section. This is a common occurrence. If you haven't messed up an IAM policy yet, you will. My next step was to apply an SCP that denied _all_ API calls. Whenever struggling with permissions in AWS, some good advice is to check CloudTrail. From the CloudTrail record, it's possible to infer what conditionals might be useful or what underlying API calls need to be allowed or disallowed.

Reviewer: ```suggestion
Surprising, right? Well, maybe not. IAM is complicated. My initial assumption was that I had somehow screwed up the Condition section. This is a common occurrence. If you haven't messed up an IAM policy yet, you will. My next step was to apply an SCP that denied _all_ API calls. 

If you're struggling with permissions in AWS, check CloudTrail. From the CloudTrail record, it's possible to infer what conditionals might be useful or what underlying API calls need to be allowed or disallowed.
```

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "DenyAll",
      "Effect": "Deny",
      "Action": [
        "*"
      ],
      "Resource": [
        "*"
      ]
    }
  ]
}
```

You had one job, SCP—don't let anything happen. Yet, this happened:

```bash
aws sts assume-role --role-arn 'arn:aws:iam::575555303850:role/TestAssumedRole' --role-session-name 'nope'
```

Credentials came back. Those credentials can't do anything, _but_ they still work.

```json
{
    "Credentials": {
        "AccessKeyId": "ASIAYMAOE7GVFJQZXUXH",
        "SecretAccessKey": "F1L82dBoYUbk9V...G1S9Yve6+GgbEiG3z5L",
        "SessionToken": "IQoJb3JpZ2luX2V...is2NDpHyKlAa",
        "Expiration": "2022-03-31T05:45:04+00:00"
    },
    "AssumedRoleUser": {
        "AssumedRoleId": "AROAYMAOE7GVP5EH66UUY:nope",
        "Arn": "arn:aws:sts::575555303850:assumed-role/TestAssumedRole/nope"
    }
}
```

> In fact, even with an SCP that denies _everything_, the AssumeRole API call succeeds and returns valid credentials. This is visible in the CloudTrail logs for the account; amongst a sea of Permission Denied errors is one successful API call: the one to AssumeRole.

Reviewer: ```suggestion
Even with an SCP that denies _everything_, the AssumeRole API call succeeds and returns valid credentials. This is visible in the CloudTrail logs for the account; amongst a sea of Permission Denied errors is one successful API call: the one to AssumeRole.
```
